k9mail-library/src/main/java/com/fsck/k9/mail/Keyword.java

```diff
@@ -0,0 +1,234 @@
+package com.fsck.k9.mail;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.lang.IndexOutOfBoundsException;
+import java.util.Random;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+import android.support.v4.graphics.ColorUtils;
+
+
+public final class Keyword extends Flag implements Parcelable {
+
+    private static final String KEYWORD_CODE_PREFIX = "KEYWORD_";
+
+    private static final ArrayList<Integer> PREDEFINED_COLORS =
+        new ArrayList<Integer>() {{
+            add(0xFFFF0000); add(0xFFFFAA00); add(0xFFFFFF00); add(0xFFAAFF00);
+            add(0xFF00FF00); add(0xFF00FFFF); add(0xFF0000FF); add(0xFFFF00FF);
+        }};
+
+
+    private static final HashMap<String, Keyword> mapExternalCodeToKeyword =
+        new HashMap<String, Keyword>();
+
+    private static final ArrayList<Keyword> orderedKeywords =
+        new ArrayList<Keyword>();
+
+
+    // User-defined keyword name for display as a tag.
+    private String name;
+
+    // Further user-defined keyword attributes
+    private boolean visible = false;
+    private int color = -1;
+
+    private static ArrayList<Integer> nextColors = null;
+
+
+    private Keyword(String code, String externalCode, String name) {
+        // The caller has verified that externalCode is valid for an IMAP
+        // keyowrd and not yet in use.
+        super(code, externalCode);
+        this.name = name;
+        color = randomColor();
+    }
+
+    public void delete() {
+        final String externalCode = this.getExternalCode();
+        if (!mapExternalCodeToKeyword.containsKey(externalCode)) {
+            throw new IllegalArgumentException(
+                "deletion of unknown IMAP keyword '" + externalCode + "'");
+        }
+        orderedKeywords.remove(this);
+        mapExternalCodeToKeyword.remove(externalCode);
+    }
+
+
+    public static Keyword getKeywordByExternalCode(String externalCode)
```

Comment:
I'm not a big fan of static methods where you're storing and loading data, they are hard to test. I'd rather these move to a `FlagsManager` class which is an application singleton.

Reply:
I have now created a `FlagManager` that manages lookup of flags and keywords, as well as dynamic creation, deletion, and order of keywords. I left the creation (and the `HashMap`s) of flags in the class `Flag`, since the flags are all hard-coded, and I did not want to make their constructor public. There are still lookup methods in `Flag`, but they no longer access `Keyword`.

On the downside, the constructor of `Keyword` and some other methods are now public, so that `FlagManager` can access them. I could have made the manager a nested class of `Keyword`, but since it manages not just keywords, but all kinds of flags, that appeared to be wrong. Also, some do-it-yourself [friend construct](https://stackoverflow.com/questions/182278/is-there-a-way-to-simulate-the-c-friend-concept-in-java?answertab=votes#tab-top) seems to be over-the-top.

Oh, and I slightly preferred `FlagManager` over `FlagsManager`, but will happily change it, if you like.  :-)